include/asterisk/res_usbradio.h

```diff
@@ -433,3 +445,67 @@ void ast_radio_time(time_t *second);
+ * of excessive audio input levels. Because CM1xxx USB audio interface ICs have an
+ * internal mixer ahead of the ADC it is not possible within the interface board
+ * analog circuitry to detect clipping at the ADC input point, thus this function
+ * enables the raw ADC data to be checked. Clipping is detected by looking for
+ * large amplitude square waves (min. 3 samples in a row > 99% FS).
+ *
+ * Data collected can be displayed from the tune-menu 'R' option or AMI
+ * "[susb/radio] tune menu-support y" function. This also shows average power levels
+ * which can be useful for optimizing audio levels and compression/limiting.
+ * In general, peak levels should be within 3-10dB of full-scale (0dBFS) and
+ * average signal power levels should be 10-20dB below full-scale.
+ *
+ * Should be passed the raw 48Ksps stereo USB frame read buffer before any filtering
+ * or downsampling has been done. Extracts the 48K mono channel and downsamples to
+ * 8Ksps (as is done in [simpleusb/usbradio]_read() but without filtering).
+ * Signal power calculation takes the square of each sample to measure RMS power.
+ * For CPU efficiency no scaling is done here. (When stats data is printed the
+ * values are scaled to dBFS.)
+ *
+ * Audio parameters of interest include:
+ * - Peak signal level over a longer time period eg. 1+ seconds (dBFS)
+ *   This defines headroom (dB) and potential for clipping
+ * - Min and max signal power levels averaged within each USB frame (dBFS)
+ *   These define average dynamic range (dB)
+ * - Min and max signal power averaged over a longer time period (dBFS)
+ *   These define total signal power and peak-to-average power ratio
+ *
+ * \author      	NR9V
+ * \param sbuf  	Rx audio sample buffer
+ * \param o	    	Rx Audio Stats data structure
+ * \param len   	Length of data in sbuf
+ * \return 	    	1 if clipping detected, 0 otherwise
+ */
+#define CLIP_LED_HOLD_TIME_MS  500
+int ast_radio_check_rx_audio(short *sbuf, struct rxaudiostatistics *o, short len);
+
+/*!
+ * \brief Display receive audio statistics.
+ *
+ * Display the audio stats buffer data in normalized units. Peak value is the largest
+ * sample value seen in the past AUDIO_STATS_LEN audio frames (1 second default).
+ * Average, min, and max signal power levels are calculated from the total signal
+ * power buffer which contains total RMS power per 20mS frame. Avg Pwr is the average
+ * of the power values in the buffer, min and max are the lowest and highest average
+ * power levels within the buffer. ClipCnt is the count of audio clipping events
+ * detected.
+ *
+ * Example output message:
+ *   RxAudioStats: Pk -2.1  Avg Pwr -32  Min -60  Max -12  dBFS  ClipCnt 0
+ *
+ * Results are scaled to double precision 0.0-1.0 and converted to log (dB)
+ * ie. 10*log10(scaledVal) for power levels.
+ *
+ * \author  		NR9V
+ * \param fd		File descriptor to print to, or if 0 print using ast_verbose()
```

Comment:
Writing to STDIN would seem to be a very non-standard use case that I'd think would never happen in app_rpt at least with this function. And looks like asterisk sets stdin and out to /dev/null/:
# pidof asterisk
787
# ls -l /proc/787/fd
total 0
lr-x------ 1 asterisk asterisk 64 Sep 15 12:11 0 -> /dev/null
l-wx------ 1 asterisk asterisk 64 Sep 15 12:11 1 -> /dev/null
Thus I think the current behavior of the function to call ast_verbose() if fd=0 is probably the best thing to do.

Reply:
I'm not objecting to using the fd==0 test.  All I was suggesting is that you have a function where you are passed a file descriptor and "0" is a valid file descriptor.  The function should not make any assumptions about whether FD 0 is STDIN or was setup for a real file.  On the other hand, you will never have an FD==-1.

Reply:
I do see the value of keeping things generic regardless of the actual usage of a function in context, so have updated that to allow 0, and print using ast_verbose only if fd < 0.

Reply:
> Thus I think the current behavior of the function to call ast_verbose() if fd=0 is probably the best thing to do.

I looked at the code and it seems you already made a change here but the comments don't line up with that, so just wanted to add that I agree with Allan, never use 0 as an indicator for lack of file descriptor, -1 is the standard value for "no FD", including in Asterisk.

> And looks like asterisk sets stdin and out to /dev/null:

I suspect that's because you're running a daemonized Asterisk, which is more typical, but STDOUT should be present when Asterisk is run in the foreground.

Reply:
I did update the function header comments to in the previous commit:
` * \param fd		File descriptor to print to, or if -1 print using ast_verbose()`

But as the function was in fact using ast_verbose for any fd < 0, I just updated that now and pushed to have that line read:
 * \param fd		File descriptor to print to, or if < 0 print using ast_verbose()